contracts/marketplace/MarketplaceV2.sol

```diff
@@ -0,0 +1,453 @@
+      nftAddress,
+      priceInWei,
+      expiresAt
+    );
+  }
+
+  /**
+    * @dev Cancel an already published order
+    *  can only be canceled by seller or the contract owner
+    * @param nftAddress - Address of the NFT registry
+    * @param assetId - ID of the published NFT
+    */
+  function _cancelOrder(address nftAddress, uint256 assetId) internal returns (Order memory) {
+    address sender = _msgSender();
+    Order memory order = orderByAssetId[nftAddress][assetId];
+
+    require(order.id != 0, "Asset not published");
+    require(order.seller == sender || sender == owner(), "Unauthorized user");
+
+    bytes32 orderId = order.id;
+    address orderSeller = order.seller;
+    address orderNftAddress = order.nftAddress;
+    delete orderByAssetId[nftAddress][assetId];
+
+    emit OrderCancelled(
+      orderId,
+      assetId,
+      orderSeller,
+      orderNftAddress
+    );
+
+    return order;
+  }
+
+  /**
+    * @dev Executes the sale for a published NFT
+    * @param nftAddress - Address of the NFT registry
+    * @param assetId - ID of the published NFT
+    * @param price - Order price
+    * @param fingerprint - Verification info for the asset
+    */
+  function _executeOrder(
+    address nftAddress,
+    uint256 assetId,
+    uint256 price,
+    bytes memory fingerprint
+  )
+   internal returns (Order memory)
+  {
+    _requireERC721(nftAddress);
+
+    address sender = _msgSender();
+
+    IERC721Verifiable nftRegistry = IERC721Verifiable(nftAddress);
+
+    if (nftRegistry.supportsInterface(InterfaceId_ValidateFingerprint)) {
+      require(
+        nftRegistry.verifyFingerprint(assetId, fingerprint),
+        "The asset fingerprint is not valid"
+      );
+    }
+    Order memory order = orderByAssetId[nftAddress][assetId];
+
+    require(order.id != 0, "Asset not published");
+
+    require(order.seller != address(0), "Invalid address");
+    require(order.seller != sender, "Unauthorized user");
+    require(order.price == price, "The price is not correct");
+    require(block.timestamp < order.expiresAt, "The order expired");
+    require(order.seller == nftRegistry.ownerOf(assetId), "The seller is no longer the owner");
+
+
+    delete orderByAssetId[nftAddress][assetId];
+
+    uint256 feesCollectorShareAmount;
+    uint256 royaltiesShareAmount;
+
+    // Fees collector share
+    if (feesCollectorCutPerMillion > 0 && feesCollector != address(0)) {
+      feesCollectorShareAmount = price.mul(feesCollectorCutPerMillion).div(1000000);
+
+      require(
+        acceptedToken.transferFrom(sender, feesCollector, feesCollectorShareAmount),
+        "Transfering the fees collector cut failed"
+      );
+    }
+
+    // Royalties share
+    if (royaltiesCutPerMillion > 0) {
+      address royaltiesReceiver = royaltiesManager.getRoyaltiesReceiver(address(nftRegistry), assetId);
+      if (royaltiesReceiver != address(0)) {
+        royaltiesShareAmount = price.mul(royaltiesCutPerMillion).div(1000000);
```

Comment:
If the user specifies that `royaltiesCutPerMillion` should be `500000`, that is, the half of the price, the operation kind of neglects the floating point part of the result.
Let's say that the price is 5, multiplying the price by 500.000 and dividing this by 1.000.000 should return `2.5` or 2 if we think of it as an integer division, and that's a bit less than what the user specified. This might be a bit confusing, as the expected value from the sale won't be what the user put before.

Reply:
Let me check if I understood it ok...

5 in your example is 5 WEI. That is a hilarious price of something. Extremely low.

 Chances are that you are talking about 5 ETH (5 * 10^18). In your math, that will be 2.5 ETH. Am I correct?